DLPX-66981 Add verbose logging to the upgrade bash scripts

sumedhbala-delphix opened the pull request:
The following change is expected to ease debugging upgrade failures by increasing the logging.
 
Modern versions of bash include a BASH_XTRACEFD specifically for specifying an alternate file descriptor for the output of set -x. The following is used to redirect the verbose logging to the syslog file on the Delphix Engine. 
"""
exec 4> >(logger -t $0)
BASH_XTRACEFD="4"
"""
The following is used to add the source file & line number to the verbose output.
"""
PS4=':${BASH_SOURCE}:${LINENO}+' :
"""

Testing: git ab-pre-push --test-upgrade-from trunk: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/2501/

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/delphix/appliance-build?pullRequest=397) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/delphix/appliance-build?pullRequest=397) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/delphix/appliance-build?pullRequest=397) it.</sub>

pzakha commented on upgrade/upgrade-scripts/common.sh:
```diff
@@ -204,3 +204,10 @@ function set_upgrade_property() {
+function turn_verbose_logging_on() {
+    exec 4> >(logger -t $0)
+    BASH_XTRACEFD="4"
+    PS4=':${BASH_SOURCE}:${LINENO}+'
```
nit: Could you add a space at the end of the prompt so that it's easier to read the messages?

prakashsurya replied:
How would you feel about changing this to:
```
    exec 4> >(logger -t "delphix-upgrade[$BASHPID]")
    BASH_XTRACEFD="4"
    PS4='${BASH_SOURCE}:${LINENO} '
```
?

This would result in message like the following:
```
Nov 20 23:06:37 ps-trunk delphix-upgrade[5772]: /var/dlpx-update/latest/upgrade:103 case "$UPGRADE_TYPE" in
Nov 20 23:06:37 ps-trunk delphix-upgrade[5772]: /var/dlpx-update/latest/upgrade:111 CONTAINER=delphix.mPIq2s4
Nov 20 23:06:37 ps-trunk delphix-upgrade[5772]: /var/dlpx-update/latest/upgrade:112 [[ -n delphix.mPIq2s4 ]]
Nov 20 23:06:37 ps-trunk delphix-upgrade[5772]: /var/dlpx-update/latest/upgrade:114 /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container start delphix.mPIq2s4
```

If we don't include the PID in the tag used with `logger`, I think it'll be difficult to discern to different processes both attempting to upgrade concurrently. Obviously this shouldn't occur, but it's not impossible (e.g. when running upgrade via the scripts), so it'd be nice to account for that.

prakashsurya replied:
When I execute an upgrade using the "latest" version, like this:
```
$ sudo /var/dlpx-update/latest/upgrade -nv deferred
```
the the path using "latest" winds up in the log as well:
```
Nov 20 23:26:45 ps-trunk delphix-upgrade[22530]: /var/dlpx-update/latest/upgrade:103 case "$UPGRADE_TYPE" in
Nov 20 23:26:45 ps-trunk delphix-upgrade[22530]: /var/dlpx-update/latest/upgrade:111 CONTAINER=delphix.Oo4cOFI
Nov 20 23:26:45 ps-trunk delphix-upgrade[22530]: /var/dlpx-update/latest/upgrade:112 [[ -n delphix.Oo4cOFI ]]
```
Ideally, I think we'd want to show the "real" path, with the upgrade image version embedded in it, rather than "latest".

It's probably not worth the effort to fix this, but I at least wanted to mention it.

pzakha commented on upgrade/upgrade-scripts/common.sh:
```diff
@@ -204,3 +204,10 @@ function set_upgrade_property() {
 		die "failed to read properties file after setting '$1=$2'"
 }
+
+function turn_verbose_logging_on() {
```
Might be useful to print a message here to inform that debug info is available in `/var/log/syslog`.

prakashsurya commented:
Looking in `/var/log/syslog`, I see this:
```
Nov 20 22:54:44 ps-trunk /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container: :/var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container:
415+for i in {1..600}
Nov 20 22:54:44 ps-trunk /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container: :/var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container:
416+machinectl status delphix.p6hXts0
Nov 20 22:54:44 ps-trunk /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container: :/var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container:
417+break
```
is this the expected format? it looks like we duplicate the path to the script.. is that intentional?

prakashsurya commented on upgrade/upgrade-scripts/common.sh:
```diff
@@ -204,3 +204,10 @@ function set_upgrade_property() {
 }
+
+function turn_verbose_logging_on() {
+    exec 4> >(logger -t $0)
```
Should we be using any specific facility name and/or log level here? i.e. via the `-p` option?

sumedhbala-delphix commented:
> Looking in `/var/log/syslog`, I see this:
> 
> ```
> Nov 20 22:54:44 ps-trunk /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container: :/var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container:
> 415+for i in {1..600}
> Nov 20 22:54:44 ps-trunk /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container: :/var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container:
> 416+machinectl status delphix.p6hXts0
> Nov 20 22:54:44 ps-trunk /var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container: :/var/dlpx-update/6.1.0.0-snapshot.20191114231533366+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3703/upgrade-container:
> 417+break
> ```
> 
> is this the expected format? it looks like we duplicate the path to the script.. is that intentional?

This is intentional, I got this from https://stackoverflow.com/questions/10813863/while-executing-shell-scripts-how-to-know-which-line-number-its-executing
"""
If multiple files are involved in running your script, you might want to include the BASH_SOURCE variable as opposed to only LINENO"""

Example: ```/var/dlpx-update/6.1.0.0-snapshot.20191106190823553+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3609/upgrade-container: ::/var/dlpx-update/6.1.0.0-snapshot.20191106190823553+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3609/common.sh:52+readlink -f /var/dlpx-update/6.1.0.0-snapshot.20191106190823553+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3609```
We can see upgrade-container calling common.sh

prakashsurya commented:
> We can see upgrade-container calling common.sh

That just means `upgrade-container` called `logger`, right? It doesn't necessarily imply `upgrade-container` was the script that "called" the line being logged. For example, if `common.sh` then called some other script, it'd still look like `upgrade-container` made that call (it'd still be the "tag" used with `logger`).

Consider another example, if `upgrade`, calls `upgrade-container`, and then `upgrade-container` calls `common.sh`, for the line executed via `common.sh` what will the "tag" be in the logs? `upgrade` or `upgrade-container`? which would we prefer?

sumedhbala-delphix commented:
> > We can see upgrade-container calling common.sh
> 
> That just means `upgrade-container` called `logger`, right? It doesn't necessarily imply `upgrade-container` was the script that "called" the line being logged. For example, if `common.sh` then called some other script, it'd still look like `upgrade-container` made that call (it'd still be the "tag" used with `logger`).
> 
> Consider another example, if `upgrade`, calls `upgrade-container`, and then `upgrade-container` calls `common.sh`, for the line executed via `common.sh` what will the "tag" be in the logs? `upgrade` or `upgrade-container`? which would we prefer?

I added the log output for PS4=':${LINENO}+' under the file just_line_no.txt
I added the log output for not using PS$ under no_PS4.txt
[Uploading just_line_no…]()
[Uploading no_PS4…]()
[just_line_no.txt](https://github.com/delphix/appliance-build/files/3957567/just_line_no.txt)
[no_PS4.txt](https://github.com/delphix/appliance-build/files/3957568/no_PS4.txt)

sumedhbala-delphix commented:
Latest testing: git ab-pre-push --test-upgrade-from trunk: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/2660/ (PASS)

Syslog snippet from the upgrade run:
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: //var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:20 get_image_path
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: //var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/common.sh:get_image_path:65 readlink -f /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:20 IMAGE_PATH=/var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:21 [[ -n /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988 ]]
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:23 CONTAINER=
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:817 [[ 0 -ne 0 ]]
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:819 case "$1" in
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:821 [[ 2 -lt 2 ]]
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:822 [[ 2 -gt 2 ]]
Jan  8 22:50:38 ip-10-110-213-214 upgrade-scripts:upgrade-container[28081]: /var/dlpx-update/6.1.0.0-snapshot.20200108202143311+jenkins-selfservice-devops-gate-master-appliance-build-master-pre-push-3988/upgrade-container:main:824 case "$2" in

sumedhbala-delphix commented:
bors r+

prakashsurya commented:
We're no longer using bors; I can merge this.
